File: MLAI/lab1-spoken-digit-recognition/lab1_spoken_digit_recognition.m
%% Lab 1 - Spoken Digit Recognition
% 语音数字识别实验
% 使用KNN和SVM算法对Audio MNIST数据集进行分类

clear; close all; clc;

%% 配置参数
data_dir = 'audio_mnist/recordings';
data_size = 2600;
test_ratio = 0.3;
n_mfcc_default = 20;

fprintf('Lab 1 - 语音数字识别\n');
fprintf(repmat('=', 1, 80)); fprintf('\n');

%% 检查数据集是否存在
if ~exist(data_dir, 'dir')
    warning('数据集目录 %s 不存在', data_dir);
    fprintf('请将Audio MNIST数据集放置在正确的目录中\n');
    return;
end

%% Step 1: 读取并分割数据集
fprintf('\nStep 1: 读取并分割数据集\n');
fprintf(repmat('-', 1, 80)); fprintf('\n');

% 获取所有.wav文件
file_info = dir(fullfile(data_dir, '*.wav'));
file_list = {file_info.name};
n_files = length(file_list);

fprintf('找到 %d 个音频文件\n', n_files);

% 随机打乱并分割数据
rng(42); % 设置随机种子
indices = randperm(n_files);
file_list = file_list(indices);

split_point = floor(n_files * test_ratio);
test_list = file_list(1:split_point);
train_list = file_list(split_point+1:end);

fprintf('训练集: %d 个样本\n', length(train_list));
fprintf('测试集: %d 个样本\n', length(test_list));

%% Step 2: 提取MFCC特征
fprintf('\nStep 2: 提取MFCC特征 (n_mfcc=%d)\n', n_mfcc_default);
fprintf(repmat('-', 1, 80)); fprintf('\n');

[train_features, train_labels] = extractMFCCFeatures(data_dir, train_list, n_mfcc_default);
[test_features, test_labels] = extractMFCCFeatures(data_dir, test_list, n_mfcc_default);

fprintf('训练集形状: [%d, %d]\n', size(train_features));
fprintf('测试集形状: [%d, %d]\n', size(test_features));

%% Step 3: 数据标准化
fprintf('\nStep 3: 数据标准化\n');
fprintf(repmat('-', 1, 80)); fprintf('\n');

% 计算均值和标准差
mu = mean(train_features, 1);
sigma = std(train_features, 0, 1);

% 标准化
train_features_std = (train_features - mu) ./ sigma;
test_features_std = (test_features - mu) ./ sigma;

% 绘制标准化后的分布
figure('Position', [100, 100, 800, 600]);
histogram(train_features_std(:, 1), 50, 'Normalization', 'pdf', ...
    'FaceAlpha', 0.6, 'DisplayName', 'Data Distribution');
hold on;
x = linspace(-3, 3, 100);
plot(x, normpdf(x, 0, 1), 'r-', 'LineWidth', 2, ...
    'DisplayName', 'Standard Normal Distribution');
xlabel('Feature Value', 'FontSize', 12);
ylabel('Density', 'FontSize', 12);
title('Standardized Feature Distribution', 'FontSize', 14);
legend('show');
grid on;
saveas(gcf, 'standardized_distribution.png');

fprintf('数据标准化完成\n');

%% Step 4: KNN分类器 (默认参数)
fprintf('\nStep 4: 训练KNN分类器 (k=5)\n');
fprintf(repmat('-', 1, 80)); fprintf('\n');

k = 5;
knn_model = fitcknn(train_features_std, train_labels, 'NumNeighbors', k);
knn_predictions = predict(knn_model, test_features_std);

% 计算性能指标
knn_accuracy = sum(knn_predictions == test_labels) / length(test_labels);
knn_f1 = calculateF1Score(test_labels, knn_predictions);

fprintf('KNN (k=%d) Accuracy: %.4f\n', k, knn_accuracy);
fprintf('KNN (k=%d) F1 Score: %.4f\n', k, knn_f1);

% 绘制混淆矩阵
plotConfusionMatrix(test_labels, knn_predictions, ...
    sprintf('KNN Confusion Matrix (k=%d)', k), 'knn_confusion_matrix.png');

%% Step 5: SVM分类器 (默认参数)
fprintf('\nStep 5: 训练SVM分类器 (kernel=rbf)\n');
fprintf(repmat('-', 1, 80)); fprintf('\n');

% 使用RBF核的多类SVM
template = templateSVM('KernelFunction', 'rbf', 'Standardize', false);
svm_model = fitcecoc(train_features_std, train_labels, 'Learners', template);
svm_predictions = predict(svm_model, test_features_std);

% 计算性能指标
svm_accuracy = sum(svm_predictions == test_labels) / length(test_labels);
svm_f1 = calculateF1Score(test_labels, svm_predictions);

fprintf('SVM (kernel=rbf) Accuracy: %.4f\n', svm_accuracy);
fprintf('SVM (kernel=rbf) F1 Score: %.4f\n', svm_f1);

% 绘制混淆矩阵
plotConfusionMatrix(test_labels, svm_predictions, ...
    'SVM Confusion Matrix (kernel=rbf)', 'svm_confusion_matrix.png');

%% 实验1: 不同MFCC特征数量对比
fprintf('\n'); fprintf(repmat('=', 1, 80)); fprintf('\n');
fprintf('实验1: 不同MFCC特征数量对比\n');
fprintf(repmat('=', 1, 80)); fprintf('\n');

mfcc_values = [10, 20, 30];
mfcc_results = [];

for i = 1:length(mfcc_values)
    n_mfcc = mfcc_values(i);
    fprintf('\n测试 n_mfcc = %d\n', n_mfcc);
    fprintf(repmat('-', 1, 60)); fprintf('\n');

    % 提取特征
    [train_feat, train_lab] = extractMFCCFeatures(data_dir, train_list, n_mfcc);
    [test_feat, test_lab] = extractMFCCFeatures(data_dir, test_list, n_mfcc);

    % 标准化
    mu_temp = mean(train_feat, 1);
    sigma_temp = std(train_feat, 0, 1);
    train_feat_std = (train_feat - mu_temp) ./ sigma_temp;
    test_feat_std = (test_feat - mu_temp) ./ sigma_temp;

    % KNN
    knn_temp = fitcknn(train_feat_std, train_lab, 'NumNeighbors', 5);
    knn_pred = predict(knn_temp, test_feat_std);
    knn_f1_temp = calculateF1Score(test_lab, knn_pred);

    % SVM
    template_temp = templateSVM('KernelFunction', 'rbf', 'Standardize', false);
    svm_temp = fitcecoc(train_feat_std, train_lab, 'Learners', template_temp);
    svm_pred = predict(svm_temp, test_feat_std);
    svm_f1_temp = calculateF1Score(test_lab, svm_pred);

    mfcc_results = [mfcc_results; n_mfcc, knn_f1_temp, svm_f1_temp];

    fprintf('KNN F1 Score: %.4f\n', knn_f1_temp);
    fprintf('SVM F1 Score: %.4f\n', svm_f1_temp);
end

% 打印对比表格
fprintf('\n'); fprintf(repmat('=', 1, 80)); fprintf('\n');
fprintf('MFCC特征数量对比结果\n');
fprintf(repmat('=', 1, 80)); fprintf('\n');
fprintf('%-15s %-20s %-20s\n', 'n_mfcc', 'KNN F1 Score', 'SVM F1 Score');
fprintf(repmat('-', 1, 80)); fprintf('\n');
for i = 1:size(mfcc_results, 1)
    fprintf('%-15d %-20.4f %-20.4f\n', mfcc_results(i, 1), ...
        mfcc_results(i, 2), mfcc_results(i, 3));
end
fprintf(repmat('=', 1, 80)); fprintf('\n');

%% 实验2: 不同KNN邻居数对比
fprintf('\n'); fprintf(repmat('=', 1, 80)); fprintf('\n');
fprintf('实验2: 不同KNN邻居数对比\n');
fprintf(repmat('=', 1, 80)); fprintf('\n');

neighbor_values = [3, 5, 7];
knn_results = [];

for i = 1:length(neighbor_values)
    k_temp = neighbor_values(i);
    fprintf('\n测试 n_neighbors = %d\n', k_temp);

    knn_temp = fitcknn(train_features_std, train_labels, 'NumNeighbors', k_temp);
    knn_pred = predict(knn_temp, test_features_std);
    f1_temp = calculateF1Score(test_labels, knn_pred);

    knn_results = [knn_results; k_temp, f1_temp];
    fprintf('F1 Score: %.4f\n', f1_temp);
end

% 打印对比表格
fprintf('\n'); fprintf(repmat('=', 1, 80)); fprintf('\n');
fprintf('KNN邻居数对比结果\n');
fprintf(repmat('=', 1, 80)); fprintf('\n');
fprintf('%-15s %-20s\n', 'n_neighbors', 'F1 Score');
fprintf(repmat('-', 1, 80)); fprintf('\n');
for i = 1:size(knn_results, 1)
    fprintf('%-15d %-20.4f\n', knn_results(i, 1), knn_results(i, 2));
end
fprintf(repmat('=', 1, 80)); fprintf('\n');

%% 实验3: 不同SVM核函数对比
fprintf('\n'); fprintf(repmat('=', 1, 80)); fprintf('\n');
fprintf('实验3: 不同SVM核函数对比\n');
fprintf(repmat('=', 1, 80)); fprintf('\n');

kernel_names = {'linear', 'rbf', 'polynomial'};
kernel_types = {'linear', 'rbf', 'polynomial'};
svm_results = cell(length(kernel_names), 2);

for i = 1:length(kernel_types)
    kernel = kernel_types{i};
    fprintf('\n测试 kernel = %s\n', kernel);

    template_temp = templateSVM('KernelFunction', kernel, 'Standardize', false);
    svm_temp = fitcecoc(train_features_std, train_labels, 'Learners', template_temp);
    svm_pred = predict(svm_temp, test_features_std);
    f1_temp = calculateF1Score(test_labels, svm_pred);

    svm_results{i, 1} = kernel_names{i};
    svm_results{i, 2} = f1_temp;

    fprintf('F1 Score: %.4f\n', f1_temp);

    % 绘制混淆矩阵
    plotConfusionMatrix(test_labels, svm_pred, ...
        sprintf('SVM Confusion Matrix (kernel=%s)', kernel), ...
        sprintf('svm_%s_confusion_matrix.png', kernel));
end

% 打印对比表格
fprintf('\n'); fprintf(repmat('=', 1, 80)); fprintf('\n');
fprintf('SVM核函数对比结果\n');
fprintf(repmat('=', 1, 80)); fprintf('\n');
fprintf('%-15s %-20s\n', 'Kernel', 'F1 Score');
fprintf(repmat('-', 1, 80)); fprintf('\n');
for i = 1:size(svm_results, 1)
    fprintf('%-15s %-20.4f\n', svm_results{i, 1}, svm_results{i, 2});
end
fprintf(repmat('=', 1, 80)); fprintf('\n');

fprintf('\n所有实验完成！\n');

%% 辅助函数

function [features, labels] = extractMFCCFeatures(data_dir, file_list, n_mfcc)
    % 从音频文件中提取MFCC特征
    %
    % 输入:
    %   data_dir: 数据目录
    %   file_list: 文件名列表
    %   n_mfcc: MFCC系数数量
    %
    % 输出:
    %   features: 特征矩阵
    %   labels: 标签向量

    n_files = length(file_list);
    labels = zeros(n_files, 1);
    features_cell = cell(n_files, 1);

    for i = 1:n_files
        file_name = file_list{i};
        file_path = fullfile(data_dir, file_name);

        try
            % 读取音频文件
            [audio_data, sample_rate] = audioread(file_path);

            % 如果是立体声，转换为单声道
            if size(audio_data, 2) > 1
                audio_data = mean(audio_data, 2);
            end

            % 提取标签
            parts = strsplit(file_name, '_');
            labels(i) = str2double(parts{1});

            % 提取MFCC特征
            mfcc_feat = mfcc(audio_data, sample_rate, 'NumCoeffs', n_mfcc);

            % 展平为向量
            features_cell{i} = mfcc_feat(:)';
        catch
            warning('处理文件 %s 时出错', file_name);
            features_cell{i} = [];
        end
    end

    % 找到所有特征的最大长度
    feature_lengths = cellfun(@length, features_cell);
    max_length = max(feature_lengths);

    % 填充或截断特征到相同长度
    features = zeros(n_files, max_length);
    for i = 1:n_files
        feat = features_cell{i};
        if ~isempty(feat)
            if length(feat) < max_length
                features(i, 1:length(feat)) = feat;
            else
                features(i, :) = feat(1:max_length);
            end
        end
    end
end

function f1 = calculateF1Score(true_labels, predicted_labels)
    % 计算微平均F1分数
    %
    % 输入:
    %   true_labels: 真实标签
    %   predicted_labels: 预测标签
    %
    % 输出:
    %   f1: F1分数

    % 微平均F1等于准确率
    f1 = sum(true_labels == predicted_labels) / length(true_labels);
end

function plotConfusionMatrix(true_labels, predicted_labels, title_text, save_path)
    % 绘制混淆矩阵
    %
    % 输入:
    %   true_labels: 真实标签
    %   predicted_labels: 预测标签
    %   title_text: 图表标题
    %   save_path: 保存路径

    % 计算混淆矩阵
    cm = confusionmat(true_labels, predicted_labels);

    % 绘图
    figure('Position', [100, 100, 800, 700]);
    imagesc(cm);
    colormap(flipud(gray));
    colorbar;

    % 添加数值标注
    [rows, cols] = size(cm);
    for i = 1:rows
        for j = 1:cols
            text(j, i, num2str(cm(i, j)), ...
                'HorizontalAlignment', 'center', ...
                'VerticalAlignment', 'middle', ...
                'Color', 'red', 'FontSize', 10);
        end
    end

    xlabel('Predicted', 'FontSize', 12);
    ylabel('Ground Truth', 'FontSize', 12);
    title(title_text, 'FontSize', 14);
    set(gca, 'XTick', 1:10, 'XTickLabel', 0:9);
    set(gca, 'YTick', 1:10, 'YTickLabel', 0:9);

    % 保存图片
    saveas(gcf, save_path);
end

File: MLAI/lab1-spoken-digit-recognition/lab1_spoken_digit_recognition.py
"""
Lab 1 - Spoken Digit Recognition
语音数字识别实验

使用KNN和SVM算法对Audio MNIST数据集进行分类
包含不同参数的性能对比实验
"""

import os
import sys
import numpy as np
from random import shuffle, seed
from scipy.io import wavfile
import matplotlib.pyplot as plt
import seaborn as sn
from scipy.stats import norm
import warnings
warnings.filterwarnings('ignore')

# 设置随机种子以确保可重复性
seed(42)
np.random.seed(42)

try:
    from librosa import feature
    from librosa.util import fix_length
except ImportError:
    print("请安装librosa: pip install librosa")
    sys.exit(1)

from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix, f1_score, classification_report
from sklearn.preprocessing import StandardScaler


class SpokenDigitRecognizer:
    """语音数字识别器类"""

    def __init__(self, data_dir="audio_mnist/recordings", data_size=2600):
        """
        初始化识别器

        参数:
            data_dir: 数据集目录路径
            data_size: 音频数据固定长度
        """
        self.data_dir = data_dir
        self.data_size = data_size
        self.train_set = None
        self.test_set = None
        self.train_label = None
        self.test_label = None
        self.scaler = StandardScaler()

    def load_and_split_data(self, test_ratio=0.3):
        """
        加载并分割数据集

        参数:
            test_ratio: 测试集比例
        """
        if not os.path.exists(self.data_dir):
            print(f"警告: 数据集目录 {self.data_dir} 不存在")
            print("请将Audio MNIST数据集放置在正确的目录中")
            return False

        # 读取文件列表
        file_list = [f for f in os.listdir(self.data_dir) if f.endswith('.wav')]
        print(f"找到 {len(file_list)} 个音频文件")

        # 分割数据集
        n = len(file_list)
        shuffle(file_list)
        split_point = int(n * test_ratio)
        test_list = file_list[:split_point]
        train_list = file_list[split_point:]

        print(f"训练集: {len(train_list)} 个样本")
        print(f"测试集: {len(test_list)} 个样本")

        return train_list, test_list

    def extract_features(self, file_list, n_mfcc=20):
        """
        从音频文件中提取MFCC特征

        参数:
            file_list: 文件名列表
            n_mfcc: MFCC系数数量

        返回:
            特征矩阵和标签数组
        """
        labels = []
        datas = []

        for file_name in file_list:
            try:
                # 读取音频文件
                file_path = os.path.join(self.data_dir, file_name)
                sample_rate, data = wavfile.read(file_path)

                # 固定长度
                data = fix_length(data, size=self.data_size)

                # 提取标签（文件名格式: {digitLabel}_{speakerName}_{index}.wav）
                label = int(file_name.split('_')[0])

                # MFCC特征提取
                data = data.astype(float)
                mfcc_features = feature.mfcc(y=data, sr=sample_rate, n_fft=1024, n_mfcc=n_mfcc)

                # 展平为一维向量
                datas.append(mfcc_features.flatten())
                labels.append(label)
            except Exception as e:
                print(f"处理文件 {file_name} 时出错: {e}")
                continue

        return np.stack(datas), np.array(labels, dtype=int)

    def prepare_data(self, n_mfcc=20, test_ratio=0.3):
        """
        准备训练和测试数据

        参数:
            n_mfcc: MFCC系数数量
            test_ratio: 测试集比例
        """
        # 加载并分割数据
        result = self.load_and_split_data(test_ratio)
        if not result:
            return False

        train_list, test_list = result

        # 提取特征
        print(f"\n提取MFCC特征 (n_mfcc={n_mfcc})...")
        self.train_set, self.train_label = self.extract_features(train_list, n_mfcc)
        self.test_set, self.test_label = self.extract_features(test_list, n_mfcc)

        print(f"训练集形状: {self.train_set.shape}")
        print(f"测试集形状: {self.test_set.shape}")

        return True

    def standardize_data(self):
        """数据标准化"""
        # 使用训练集拟合scaler
        self.scaler.fit(self.train_set)

        # 转换训练集和测试集
        self.train_set = self.scaler.transform(self.train_set)
        self.test_set = self.scaler.transform(self.test_set)

        print("数据标准化完成")

    def plot_standardized_distribution(self, save_path=None):
        """绘制标准化后的数据分布"""
        fig, ax = plt.subplots(figsize=(10, 6))

        # 绘制第一个特征的分布
        ax.hist(self.train_set[:, 0], bins=50, density=True, alpha=0.6, label="Data Distribution")

        # 绘制标准正态分布
        x = np.arange(-3, 3, 0.01)
        ax.plot(x, norm.pdf(x, 0, 1.), 'r-', linewidth=2, label="Standard Normal Distribution")

        ax.set_xlabel("Feature Value", fontsize=12)
        ax.set_ylabel("Density", fontsize=12)
        ax.set_title("Standardized Feature Distribution", fontsize=14)
        ax.legend()
        ax.grid(True, alpha=0.3)

        if save_path:
            plt.savefig(save_path, dpi=300, bbox_inches='tight')
        plt.show()

    def train_and_evaluate_knn(self, n_neighbors=5):
        """
        训练并评估KNN模型

        参数:
            n_neighbors: 邻居数量

        返回:
            模型、预测结果、F1分数
        """
        print(f"\n训练KNN模型 (n_neighbors={n_neighbors})...")

        # 创建并训练模型
        knn = KNeighborsClassifier(n_neighbors=n_neighbors, metric="minkowski")
        knn.fit(self.train_set, self.train_label)

        # 预测
        predictions = knn.predict(self.test_set)

        # 评估
        f1 = f1_score(self.test_label, predictions, average="micro")

        print(f"KNN (k={n_neighbors}) F1 Score: {f1:.4f}")

        return knn, predictions, f1

    def train_and_evaluate_svm(self, kernel='rbf', gamma='auto'):
        """
        训练并评估SVM模型

        参数:
            kernel: 核函数类型
            gamma: 核函数系数

        返回:
            模型、预测结果、F1分数
        """
        print(f"\n训练SVM模型 (kernel={kernel})...")

        # 创建并训练模型
        svm = SVC(gamma=gamma, kernel=kernel)
        svm.fit(self.train_set, self.train_label)

        # 预测
        predictions = svm.predict(self.test_set)

        # 评估
        f1 = f1_score(self.test_label, predictions, average="micro")

        print(f"SVM (kernel={kernel}) F1 Score: {f1:.4f}")

        return svm, predictions, f1

    def plot_confusion_matrix(self, predictions, title="Confusion Matrix", save_path=None):
        """
        绘制混淆矩阵

        参数:
            predictions: 预测结果
            title: 图表标题
            save_path: 保存路径
        """
        # 计算混淆矩阵
        cm = confusion_matrix(self.test_label, predictions)

        # 绘图
        plt.figure(figsize=(10, 8))
        sn.heatmap(cm, annot=True, fmt='d', cmap='Blues', cbar=True,
                   xticklabels=range(10), yticklabels=range(10))
        plt.title(title, fontsize=14)
        plt.ylabel("Ground Truth", fontsize=12)
        plt.xlabel("Predicted", fontsize=12)

        if save_path:
            plt.savefig(save_path, dpi=300, bbox_inches='tight')
        plt.show()


def experiment_mfcc_comparison():
    """
    实验1: 比较不同MFCC特征数量的性能
    """
    print("=" * 80)
    print("实验1: 不同MFCC特征数量对比")
    print("=" * 80)

    mfcc_values = [10, 20, 30]
    results = []

    for n_mfcc in mfcc_values:
        print(f"\n{'='*60}")
        print(f"测试 n_mfcc = {n_mfcc}")
        print(f"{'='*60}")

        recognizer = SpokenDigitRecognizer()

        if not recognizer.prepare_data(n_mfcc=n_mfcc):
            print("数据准备失败，跳过此实验")
            continue

        recognizer.standardize_data()

        # 使用固定参数测试KNN和SVM
        _, _, knn_f1 = recognizer.train_and_evaluate_knn(n_neighbors=5)
        _, _, svm_f1 = recognizer.train_and_evaluate_svm(kernel='rbf')

        results.append({
            'n_mfcc': n_mfcc,
            'KNN_F1': knn_f1,
            'SVM_F1': svm_f1
        })

    # 打印对比表格
    print("\n" + "=" * 80)
    print("MFCC特征数量对比结果")
    print("=" * 80)
    print(f"{'n_mfcc':<15} {'KNN F1 Score':<20} {'SVM F1 Score':<20}")
    print("-" * 80)
    for r in results:
        print(f"{r['n_mfcc']:<15} {r['KNN_F1']:<20.4f} {r['SVM_F1']:<20.4f}")
    print("=" * 80)

    return results


def experiment_knn_neighbors():
    """
    实验2: 比较不同KNN邻居数的性能
    """
    print("\n" + "=" * 80)
    print("实验2: 不同KNN邻居数对比")
    print("=" * 80)

    neighbor_values = [3, 5, 7]
    results = []

    # 准备数据（使用n_mfcc=20）
    recognizer = SpokenDigitRecognizer()
    if not recognizer.prepare_data(n_mfcc=20):
        print("数据准备失败")
        return None

    recognizer.standardize_data()

    for n_neighbors in neighbor_values:
        _, predictions, f1 = recognizer.train_and_evaluate_knn(n_neighbors=n_neighbors)

        results.append({
            'n_neighbors': n_neighbors,
            'F1_Score': f1
        })

    # 打印对比表格
    print("\n" + "=" * 80)
    print("KNN邻居数对比结果")
    print("=" * 80)
    print(f"{'n_neighbors':<15} {'F1 Score':<20}")
    print("-" * 80)
    for r in results:
        print(f"{r['n_neighbors']:<15} {r['F1_Score']:<20.4f}")
    print("=" * 80)

    return results


def experiment_svm_kernels():
    """
    实验3: 比较不同SVM核函数的性能
    """
    print("\n" + "=" * 80)
    print("实验3: 不同SVM核函数对比")
    print("=" * 80)

    kernels = ['linear', 'rbf', 'poly']
    results = []

    # 准备数据（使用n_mfcc=20）
    recognizer = SpokenDigitRecognizer()
    if not recognizer.prepare_data(n_mfcc=20):
        print("数据准备失败")
        return None

    recognizer.standardize_data()

    for kernel in kernels:
        _, predictions, f1 = recognizer.train_and_evaluate_svm(kernel=kernel)

        # 绘制混淆矩阵
        recognizer.plot_confusion_matrix(
            predictions,
            title=f"SVM Confusion Matrix (kernel={kernel})",
            save_path=f"svm_{kernel}_confusion_matrix.png"
        )

        results.append({
            'kernel': kernel,
            'F1_Score': f1
        })

    # 打印对比表格
    print("\n" + "=" * 80)
    print("SVM核函数对比结果")
    print("=" * 80)
    print(f"{'Kernel':<15} {'F1 Score':<20}")
    print("-" * 80)
    for r in results:
        print(f"{r['kernel']:<15} {r['F1_Score']:<20.4f}")
    print("=" * 80)

    return results


def main():
    """主函数"""
    print("Lab 1 - 语音数字识别")
    print("=" * 80)

    # 基础实验：使用默认参数
    print("\n基础实验：使用默认参数 (n_mfcc=20, n_neighbors=5, kernel=rbf)")
    print("=" * 80)

    recognizer = SpokenDigitRecognizer()

    # 准备数据
    if not recognizer.prepare_data(n_mfcc=20):
        print("数据集未找到，请确保Audio MNIST数据集在正确的位置")
        print("程序将继续但无法运行实验")
        return

    # 标准化
    recognizer.standardize_data()

    # 绘制标准化分布
    recognizer.plot_standardized_distribution(save_path="standardized_distribution.png")

    # 训练和评估KNN
    knn_model, knn_predictions, knn_f1 = recognizer.train_and_evaluate_knn(n_neighbors=5)
    recognizer.plot_confusion_matrix(
        knn_predictions,
        title="KNN Confusion Matrix (k=5)",
        save_path="knn_confusion_matrix.png"
    )

    # 训练和评估SVM
    svm_model, svm_predictions, svm_f1 = recognizer.train_and_evaluate_svm(kernel='rbf')
    recognizer.plot_confusion_matrix(
        svm_predictions,
        title="SVM Confusion Matrix (kernel=rbf)",
        save_path="svm_confusion_matrix.png"
    )

    # 参数对比实验
    print("\n\n开始参数对比实验...")

    # 实验1: MFCC特征数量对比
    mfcc_results = experiment_mfcc_comparison()

    # 实验2: KNN邻居数对比
    knn_results = experiment_knn_neighbors()

    # 实验3: SVM核函数对比
    svm_results = experiment_svm_kernels()

    print("\n所有实验完成！")


if __name__ == "__main__":
    main()
